platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,585 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include "config.h"
+
+#ifdef ODP_MDEV
+
+#include <odp_posix_extensions.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <linux/vfio.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <dirent.h>
+
+#include <config.h>
+#include <odp_posix_extensions.h>
+#include <odp/api/hints.h>
+
+#include <pktio/mdev.h>
+#include <pktio/uapi_net_mdev.h>
+
+#include <odp_debug_internal.h>
+#include <odp_align_internal.h>
+
+/**
+ * returns a valid VFIO container
+ * fd must be closed by caller
+ */
+static int get_container(void)
+{
+	int ret;
+	int container;
+	/* Create a new container */
+	container = open("/dev/vfio/vfio", O_RDWR);
+
+	if (container < 0)
+		return container;
+
+	ret = ioctl(container, VFIO_GET_API_VERSION);
+	if (ret != VFIO_API_VERSION) {
+		ODP_ERR("Unknown API version\n");
+		goto out;
+	}
+
+	if (!ioctl(container, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU)) {
+		ODP_ERR("Doesn't support the IOMMU driver we want\n");
+		goto out;
+	}
+
+	return container;
+out:
+	close(container);
+	container = -1;
+	return ret;
```

Comment:
"ret" is arbitrary, zero to indicate success and non-zero (either negative or positive) to indicate failures?

Reply:
Thanks, fixed